Assignment/palindrome_test.c

```diff
@@ -0,0 +1,128 @@
+
+/// <summary>To check if the input number is valid</summary>
+int IsValid (const char* num) {
+   if (*num == '\0' || num == NULL) return 0;
+   if (*num == '-') num++;
+   while (*num) {
+      if (!isdigit ((unsigned char)*num)) return 0;
+      num++;
+   }
+   return 1;
+}
+
+void TestCases () {
+   char* inputs[] = { "vini", "I did,did I", "Ma*)d(*am", "@@@##!!!", "Mam!",
+                      "Was it a car or a cat I saw?", "Don't nod", "Hello",
+                      "4334", "2147447412", "34543", "-121", "*-23" ,"?@343", "6", " " },
+      * expOutputs[] = { "Not a Palindrome", "Palindrome", "Palindrome", "Invalid Input",
+                             "Palindrome", "Palindrome", "Palindrome", "Not a Palindrome",
+                             "Palindrome", "Palindrome", "Palindrome", "Not a Palindrome",
+                             "Invalid Input", "Invalid Input", "Palindrome", "Invalid Input" };
+   int numOfTestCases = sizeof (inputs) / sizeof (inputs[0]);
+   printf ("\n-------Testcases for String and Integer Input------\n\n"
+           "| %-30s | %-30s | %-20s | %-20s | %-10s |\n"
+           "|--------------------------------|--------------------------------|"
+           "----------------------|----------------------|------------------|\n",
+           "Input", "Reverse", "Expected Output", "Actual Output", "Test Case Result");
+   for (int i = 0; i < numOfTestCases; i++) {
+      char* input = inputs[i];
+      char reversedInput[100] = { 0 };
+      int len = strlen (input), reversed = 0, num = atoi (inputs[i]),
+         actualOutput = IsValid (input) ? ReverseNumber (num, &reversed), (num == reversed ? 1 : 0)
+         : PalindromeCheck (input);
+      for (int j = 0; j < len; j++) reversedInput[j] = input[len - j - 1];
+      reversedInput[len] = '\0';
+      const char* isPal = actualOutput == 1 ? "Palindrome" : !actualOutput ?
+         "Not a Palindrome" : "Invalid Input",
+         * res = strcmp (isPal, expOutputs[i]) == 0 ?
+         GREEN_TEXT "Pass" RESET_TEXT : RED_TEXT "Fail" RESET_TEXT;
+      printf ("| %-30s | %-30s | %-20s | %-20s | %-27s |\n", input, reversedInput, expOutputs[i],
+              isPal, res);
+   }
+   printf ("\n\n");
+}
+
+void UserInput () {
+   char input[100];
+   printf ("\nEnter a word or number: ");
+   fgets (input, sizeof (input), stdin);
+   input[strcspn (input, "\n")] = '\0';
+   if (IsValid (input)) {
+      char* endPtr;
+      errno = 0;
+      long num = strtol (input, &endPtr, 10);
+      if (endPtr == input || *endPtr != '\0') printf ("Invalid Input\n");
+      else if (errno == ERANGE || num > INT_MAX || num < INT_MIN)
+         printf (RED_TEXT "Overflow: The number is out of the valid range.\n" RESET_TEXT);
+      else {
+         if (num < 0) printf ("Not a Palindrome\n");
+         else {
+            int reversed = 0;
+            ReverseNumber ((int)num, &reversed);
+            printf ("Reversed Number: %d\n%s\n", reversed, ((reversed > INT_MAX || reversed < 0)

```

Comment:
Overflow has nothing to do with reversed < 0, moreover if num < 0 then reversed will also be < 0, so this check does nothing.
```suggestion
            printf ("Reversed Number: %d\n%s\n", reversed,
       reversed > INT_MAX ? "Overflow" : num == reversed ? "Palindrome" : "Not a Palindrome");

```